packages/@pollyjs/adapter-node-http/src/utils/transport-wrapper.js

```diff
@@ -0,0 +1,212 @@
+import http from 'http';
+import zlib from 'zlib';
+
+import { URL } from '@pollyjs/utils';
+
+const NATIVE_END = Symbol();
+const LISTENERS = Symbol();
+
+const nativeRequestMapping = new WeakMap();
+
+export default class TransportWrapper {
+  constructor(transport, { name, adapter }) {
+    this.name = name;
+    this.adapter = adapter;
+    this.transport = transport;
+  }
+
+  isPatched() {
+    return nativeRequestMapping.has(this.transport);
+  }
+
+  patch() {
+    // make sure it's not already patched
+    this.adapter.assert(
+      `The ${
+        this.name
+      } transport has already been patched, please stop any running Polly instances`,
+      !this.isPatched()
+    );
+
+    nativeRequestMapping.set(this.transport, this.transport.request);
+    this.transport.request = this.createRequestWrapper();
+  }
+
+  restore() {
+    this.adapter.assert(
+      `Cannot restore unpatched transport ${this.name}`,
+      this.isPatched()
+    );
+
+    this.transport.request = nativeRequestMapping.get(this.transport);
+    nativeRequestMapping.delete(this.transport);
+  }
+
+  getHttpResponseData(res) {
+    return new Promise(resolve => {
+      let stream;
+
+      switch (res.headers['content-encoding']) {
+        case 'gzip':
+        case 'compress':
+        case 'deflate':
+          stream = res.pipe(zlib.createUnzip());
+          delete res.headers['content-encoding'];
+          break;
+        default:
+          stream = res;
+      }
+
+      const resBuffer = [];
+
+      stream.on('data', chunk => resBuffer.push(chunk));
+
+      stream.on('end', () => {
+        const data = Buffer.concat(resBuffer).toString('utf8');
+
+        resolve(data);
+      });
+    });
+  }
+
+  async passthrough(req) {
+    const res = await new Promise(resolve => {
```

Comment:
I just realized that the way you are handling the passthrough doesnt match how we handle it in the other adapters which I believe will complicate this a bit more. What needs to happen here is the passthrough should create a completely new request and take the headers, url, method, etc from the PollyRequest instance since they can be modified before the request gets sent out. 

Consider the following scenario:

```js
const { server } = polly;

// Modify the authorization request headers before it gets sent out
server.get('/foo').on('request', req => req.setHeader('Authorization', 'abc123'));

// Create my request
const req = http.get('/foo', { headers: { authorization: 'my-auth-token' } });

// After this gets called, the actual passed-through request should contain the modified header
req.end();
```

A good example for how we do this is in the `adapter-xhr`. We create a fake XHR object that never actually makes the request. When `xhr.send()` is called, we create a brand new real XHR instance and send all the data we have present in the PollyRequest.

Reply:
I saw your implementation, i understand now. I was afraid to break the original request but as long as we bind user's event handlers we're ok 😄